File: src/main/java/pl/swmud/ns/swaedit/map/Mapper.java
package pl.swmud.ns.swaedit.map;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;

import javax.media.opengl.GLException;

import pl.swmud.ns.swaedit.gui.MapWidget;
import pl.swmud.ns.swmud._1_0.area.Area;
import pl.swmud.ns.swmud._1_0.area.Rooms.Room;
import pl.swmud.ns.swmud._1_0.area.Rooms.Room.Exits.Exit;

public class Mapper {
	private Area area;
	private int islandNo;
	private SortedMap<BigInteger, MapRoom> mapRooms = new TreeMap<BigInteger, MapRoom>();
	private SortedMap<Integer, List<MapRoom>> islandRooms = new TreeMap<Integer, List<MapRoom>>();
	private int[] layers;

	public Mapper(Area area) {
		this.area = area;
	}

	public MapWidget makeMap() {
		for (Room room : area.getRooms().getRoom()) {
			if (!isAlreadyMade(room)) {
				RoomCoords coords = new RoomCoords(0, 0, 0, islandNo++);
				// System.out.println("making room: " + room.getVnum());
				MapRoom mr = new MapRoom(room, coords);
				mapRooms.put(room.getVnum(), mr);
				makeMapRoom(mr, room);
			}
		}

		assignRevExits();

		createIslands();
		
		checkCoords();

		MapWidget mw = null;
		try {
			mw = new MapWidget(islandRooms, 0, islandNo, false);
			mw.show(0, 0);
        } catch (GLException e) {
        	if (mw != null) {
            	mw.close();
            }
        	try {
            	mw = new MapWidget(islandRooms, 0, islandNo, true);
    			mw.show(0, 0);
            } catch (GLException e1) {
            	e1.printStackTrace();
            	if (mw != null) {
                	mw.close();
                }
            	mw = null;
            }
        }
		return mw;
	}

	private void makeMapRoom(MapRoom parent, Room room) {
		for (Exit exit : room.getExits().getExit()) {
			Room childRoom = findRoom(exit);
			/* omit loop exits */
			if (childRoom != null && !childRoom.getVnum().equals(room.getVnum())) {
				// System.out.println("found child room: " +
				// childRoom.getVnum());
				ExitWrapper exWrapper = new ExitWrapper(exit);
				MapRoom mr = new MapRoom(childRoom, getCoords(exit, parent.getCoords()), exWrapper);
				parent.addChildRoom(mr, exWrapper);
				if (!isAlreadyMade(childRoom)) {
					mapRooms.put(childRoom.getVnum(), mr);
					makeMapRoom(mr, childRoom);
				}
			}
		}
	}

	private void checkCoords() {
		Map<RoomCoords, MapRoom> coordsRooms = new HashMap<RoomCoords, MapRoom>();
		layers = new int[islandNo];

		for (MapRoom attempted : mapRooms.values()) {
			MapRoom current = coordsRooms.get(attempted.getCoords()); 
			if (current != null) {
//				System.out.println("replace attempt, current: "+current.getRoom().getVnum() + ", attempted: "+attempted.getRoom().getVnum());
				RoomCoords rc = attempted.getCoords(); 
				rc.setLayer(++layers[rc.getIslandNo()]);
			}
			coordsRooms.put(attempted.getCoords(), attempted);
		}
	}

	private void assignRevExits() {
		for (MapRoom mr : mapRooms.values()) {
			for (ExitWrapper exit : mr.getMapRooms().keySet()) {
				MapRoom toRoom = mapRooms.get(exit.getVnum());
				if (toRoom != null && exit.getRevExit() == null) {
					exit.setRevExit(getRevExit(exit, mr, toRoom));
					if (exit.getRevExit() != null) {
//						System.out.println("rev exit set to: " + mr.getRoom().getVnum() + ","
//						        + exit.getRevExit().getVnum());
					}
				}
				
				/* set distnace */
				if (mr.getDistance(toRoom) > 1) {
	                exit.setDistant();
                }
			}
		}
	}

	private ExitWrapper getRevExit(Exit exit, MapRoom fromRoom, MapRoom toRoom) {
		for (ExitWrapper revExit : toRoom.getMapRooms().keySet()) {
			if (fromRoom.getRoom().getVnum().equals(revExit.getVnum())
			        && exit.getDirection() == revExit.getRevDirection()) {
				return revExit;
			}
		}

		return null;
	}

	private boolean isAlreadyMade(Room room) {
		return mapRooms.get(room.getVnum()) != null;
	}

	private Room findRoom(Exit exit) {
		for (Room room : area.getRooms().getRoom()) {
			if (room.getVnum().equals(exit.getVnum())) {
				return room;
			}
		}

		return null;
	}

	private RoomCoords getCoords(Exit exit, RoomCoords crds) {
		if (exit == null) {
			System.err.println("exit==null");
			return new RoomCoords(-1, -1, 0, islandNo);
		}

		RoomCoords coords = (RoomCoords) crds.clone();
		switch (exit.getDirection()) {
		case 0: /* north */
			coords.setZ(coords.getZ() - 1);
			break;

		case 1: /* east */
			coords.setX(coords.getX() + 1);
			break;

		case 2: /* south */
			coords.setZ(coords.getZ() + 1);
			break;

		case 3: /* west */
			coords.setX(coords.getX() - 1);
			break;

		case 4: /* up */
			coords.setY(coords.getY() + 1);
			break;

		case 5: /* down */
			coords.setY(coords.getY() - 1);
			break;

		case 6: /* north-east */
			coords.setZ(coords.getZ() - 1);
			coords.setX(coords.getX() + 1);
			break;

		case 7: /* north-west */
			coords.setZ(coords.getZ() - 1);
			coords.setX(coords.getX() - 1);
			break;

		case 8: /* south-east */
			coords.setZ(coords.getZ() + 1);
			coords.setX(coords.getX() + 1);
			break;

		case 9: /* south-west */
			coords.setZ(coords.getZ() + 1);
			coords.setX(coords.getX() - 1);
			break;

		default: /* somewhere: 10 */
			System.err.println("virtual exits not yet supported!");
			/* FIXME: find a way to mark that the exit is virtual */
		}

		return coords;
	}

	private void createIslands() {
		for (int i = 0; i < islandNo; ++i) {
			List<MapRoom> iRooms = new LinkedList<MapRoom>();
			for (MapRoom mr : mapRooms.values()) {
				if (mr.getCoords().getIslandNo() == i) {
					iRooms.add(mr);
				}
			}
			islandRooms.put(i, iRooms);
		}
	}

	public int getWidth(int islandNo) {
		return new RoomSpread(islandRooms.get(islandNo)) {
			@Override
			public int getValue(RoomCoords coords) {
				return coords.getX();
			}
		}.getMiddle();
	}

	public int getHeight(int islandNo) {
		return new RoomSpread(islandRooms.get(islandNo)) {
			@Override
			public int getValue(RoomCoords coords) {
				return coords.getY();
			}
		}.getMiddle();
	}

	public int getDepth(int islandNo) {
		return new RoomSpread(islandRooms.get(islandNo)) {
			@Override
			public int getValue(RoomCoords coords) {
				return coords.getZ();
			}
		}.getMiddle();
	}
}

File: src/map/Mapper.cpp
#include "map/Mapper.h"
#include "map/MapWidget.h"
#include "map/RoomSpread.h"
#include "model/Area.h"

#include <QDebug>

Mapper::Mapper(Area *area)
    : area_(area)
{
}

Mapper::~Mapper()
{
    cleanup();
}

void Mapper::cleanup()
{
    qDeleteAll(allExitWrappers_);
    allExitWrappers_.clear();
    qDeleteAll(allMapRooms_);
    allMapRooms_.clear();
    mapRooms_.clear();
    islandRooms_.clear();
    layers_.clear();
    islandNo_ = 0;
}

void Mapper::createMap()
{
    cleanup();

    for (int i = 0; i < area_->rooms.size(); ++i) {
        Room *room = &area_->rooms[i];
        if (!isAlreadyMade(room)) {
            RoomCoords coords(0, 0, 0, islandNo_++);
            MapRoom *mr = new MapRoom(room, coords);
            allMapRooms_.append(mr);
            mapRooms_.insert(room->vnum, mr);
            makeMapRoom(mr, room);
        }
    }

    assignRevExits();
    createIslands();
    checkCoords();
}

MapWidget *Mapper::makeMap()
{
    createMap();

    MapWidget *mw = new MapWidget(islandRooms_, 0, islandNo_);
    // Transfer ownership of map data to the widget
    QList<MapRoom *> rooms;
    QList<ExitWrapper *> exits;
    releaseOwnership(rooms, exits);
    mw->takeOwnership(rooms, exits);
    return mw;
}

void Mapper::makeMap(MapWidget *mw)
{
    createMap();
    mw->refreshMap(islandRooms_, islandNo_);
    // Transfer ownership of new map data to the widget
    QList<MapRoom *> rooms;
    QList<ExitWrapper *> exits;
    releaseOwnership(rooms, exits);
    mw->takeOwnership(rooms, exits);
}

void Mapper::releaseOwnership(QList<MapRoom *> &outMapRooms, QList<ExitWrapper *> &outExitWrappers)
{
    outMapRooms = allMapRooms_;
    outExitWrappers = allExitWrappers_;
    allMapRooms_.clear();
    allExitWrappers_.clear();
}

void Mapper::makeMapRoom(MapRoom *parent, Room *room)
{
    for (int i = 0; i < room->exits.size(); ++i) {
        const Exit &exit = room->exits[i];
        Room *childRoom = findRoom(exit);
        /* omit loop exits */
        if (childRoom != nullptr && childRoom->vnum != room->vnum) {
            ExitWrapper *exWrapper = new ExitWrapper(exit);
            allExitWrappers_.append(exWrapper);
            MapRoom *mr = new MapRoom(childRoom, getCoords(exit, parent->coords()), exWrapper);
            allMapRooms_.append(mr);
            parent->addChildRoom(mr, exWrapper);
            if (!isAlreadyMade(childRoom)) {
                mapRooms_.insert(childRoom->vnum, mr);
                makeMapRoom(mr, childRoom);
            }
        }
    }
}

void Mapper::checkCoords()
{
    QHash<RoomCoords, MapRoom *> coordsRooms;
    layers_.clear();
    layers_.resize(islandNo_);
    for (int i = 0; i < islandNo_; ++i)
        layers_[i] = 0;

    QMapIterator<int, MapRoom *> it(mapRooms_);
    while (it.hasNext()) {
        it.next();
        MapRoom *attempted = it.value();
        MapRoom *current = coordsRooms.value(attempted->coords(), nullptr);
        if (current != nullptr) {
            RoomCoords &rc = attempted->coords();
            rc.setLayer(++layers_[rc.islandNo()]);
        }
        coordsRooms.insert(attempted->coords(), attempted);
    }
}

void Mapper::assignRevExits()
{
    QMapIterator<int, MapRoom *> it(mapRooms_);
    while (it.hasNext()) {
        it.next();
        MapRoom *mr = it.value();
        QMapIterator<ExitWrapper *, MapRoom *> exitIt(mr->mapRooms());
        while (exitIt.hasNext()) {
            exitIt.next();
            ExitWrapper *exit = exitIt.key();
            MapRoom *toRoom = mapRooms_.value(exit->vnum(), nullptr);
            if (toRoom != nullptr && exit->revExit() == nullptr) {
                exit->setRevExit(getRevExit(
                    Exit{exit->direction(), exit->description(), exit->keyword(),
                         exit->flags(), exit->key(), exit->vnum(), exit->distance()},
                    mr, toRoom));
            }

            /* set distance */
            if (toRoom != nullptr && mr->getDistance(*toRoom) > 1) {
                exit->setDistant();
            }
        }
    }
}

ExitWrapper *Mapper::getRevExit(const Exit &exit, MapRoom *fromRoom, MapRoom *toRoom)
{
    QMapIterator<ExitWrapper *, MapRoom *> it(toRoom->mapRooms());
    while (it.hasNext()) {
        it.next();
        ExitWrapper *revExit = it.key();
        if (fromRoom->room()->vnum == revExit->vnum()
            && exit.direction == revExit->getRevDirection()) {
            return revExit;
        }
    }
    return nullptr;
}

bool Mapper::isAlreadyMade(Room *room) const
{
    return mapRooms_.contains(room->vnum);
}

Room *Mapper::findRoom(const Exit &exit) const
{
    for (int i = 0; i < area_->rooms.size(); ++i) {
        if (area_->rooms[i].vnum == exit.vnum) {
            return const_cast<Room *>(&area_->rooms[i]);
        }
    }
    return nullptr;
}

RoomCoords Mapper::getCoords(const Exit &exit, const RoomCoords &crds) const
{
    RoomCoords coords = crds.clone();
    switch (exit.direction) {
    case 0: /* north */
        coords.setZ(coords.z() - 1);
        break;
    case 1: /* east */
        coords.setX(coords.x() + 1);
        break;
    case 2: /* south */
        coords.setZ(coords.z() + 1);
        break;
    case 3: /* west */
        coords.setX(coords.x() - 1);
        break;
    case 4: /* up */
        coords.setY(coords.y() + 1);
        break;
    case 5: /* down */
        coords.setY(coords.y() - 1);
        break;
    case 6: /* north-east */
        coords.setZ(coords.z() - 1);
        coords.setX(coords.x() + 1);
        break;
    case 7: /* north-west */
        coords.setZ(coords.z() - 1);
        coords.setX(coords.x() - 1);
        break;
    case 8: /* south-east */
        coords.setZ(coords.z() + 1);
        coords.setX(coords.x() + 1);
        break;
    case 9: /* south-west */
        coords.setZ(coords.z() + 1);
        coords.setX(coords.x() - 1);
        break;
    default: /* somewhere: 10 */
        qWarning() << "virtual exits not yet supported!";
        break;
    }
    return coords;
}

void Mapper::createIslands()
{
    for (int i = 0; i < islandNo_; ++i) {
        QList<MapRoom *> iRooms;
        QMapIterator<int, MapRoom *> it(mapRooms_);
        while (it.hasNext()) {
            it.next();
            if (it.value()->coords().islandNo() == i) {
                iRooms.append(it.value());
            }
        }
        islandRooms_.insert(i, iRooms);
    }
}

int Mapper::getWidth(int islandNo) const
{
    if (!islandRooms_.contains(islandNo))
        return 0;
    return getXMiddle(islandRooms_.value(islandNo));
}

int Mapper::getHeight(int islandNo) const
{
    if (!islandRooms_.contains(islandNo))
        return 0;
    return getYMiddle(islandRooms_.value(islandNo));
}

int Mapper::getDepth(int islandNo) const
{
    if (!islandRooms_.contains(islandNo))
        return 0;
    return getZMiddle(islandRooms_.value(islandNo));
}
